dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/IconController.java

```diff
@@ -126,17 +128,58 @@ public void getIconData(@PathVariable String key, HttpServletResponse response)
   @GetMapping
   public @ResponseBody List<IconResponse> getAllIcons(
       @RequestParam(required = false) String[] keywords) {
+
+    IconCriteria iconCriteria = IconCriteria.of(0, 0, true);
+
     List<Icon> icons;
 
     if (keywords == null) {
-      icons = iconService.getIcons();
+      icons = iconService.getIcons(iconCriteria);
     } else {
       icons = iconService.getIcons(keywords);
     }
 
     return icons.stream().map(iconMapper::from).toList();
   }
 
+  @GetMapping(value = "/custom")
+  public @ResponseBody PagingWrapper<IconResponse> getAllCustomIcons(
+      @RequestParam(defaultValue = "1") int page,
+      @RequestParam(defaultValue = "5") int size,
+      @RequestParam(defaultValue = "false") boolean skipPaging,
+      @RequestParam(required = false) String[] keywords) {
+
+    IconCriteria iconCriteria = IconCriteria.of(page, size, skipPaging);
+
+    List<CustomIcon> icons;
+
+    if (keywords == null) {
+      icons = iconService.getCustomIcons(iconCriteria);
+    } else {
+      icons = iconService.getCustomIcons(keywords);
```

Comment:
If there are keywords then there is no pagination?
I am not sure why this distinction is made here.
In my mind, keywords would be a field in `IconCriteria` and you would apply this distinction only in the store where you will construct the query with or without the filter on keywords.
This way, you will also have pagination in any case

Reply:
when keywords are provided then response will not contains lot of Icons. It is kind of filter that is applied in the query to filter out Icons.